store/smt.go

```diff
@@ -419,6 +428,165 @@ func (s *SMT) validateTarget() lib.ErrorI {
+			return nil, err
+		}
+
+		proof = append(proof, &lib.Node{
+			Key:     node.Key.bytes(),
+			Value:   node.Value,
+			Bitmask: int32(order),
+		})
+		proof = append(proof, &lib.Node{
+			Key:     siblingNode.Key.bytes(),
+			Value:   siblingNode.Value,
+			Bitmask: int32(order),
+		})
+
+		// If the proof slice contains only two nodes, it signifies that it includes only
+		// the leaf nodes (or the parents of the leaf in the case of non-membership).
+		// Verify the sibling's position to ascertain whether it is a left or right child.
+		// If it is a left child, the two node positions need to be swapped to construct
+		// the proof accurately.
+		if len(proof) == 2 && order == leftChild {
+			proof[0], proof[1] = proof[1], proof[0]
+		}
+	}
+
+	return proof, nil
+}
+
+// VerifyProof verifies a Sparse Merkle Tree proof for a given value
+// reconstructing the root hash and comparing it against the provided root hash
+// depending on the proof type (membership or non-membership)
+func (s *SMT) VerifyProof(key []byte, value []byte, validateMembership bool, root []byte, proof []*lib.Node) (bool, lib.ErrorI) {
+	// A valid proof is expected to have at least two nodes (the leaf nodes) in order to build the root
+	// and always should include an even number of nodes representing the siblings of the nodes in the tree
+	if len(proof) < 2 && len(proof)%2 != 0 {
+		return false, ErrInvalidMerkleTreeProof()
+	}
+
+	// Generate the inital hash from the leaves of the proof
+	leftLeaf := proof[0]
+	rightLeaf := proof[1]
+
+	hash := crypto.Hash(append(append(leftLeaf.Key, leftLeaf.Value...),
+		append(rightLeaf.Key, rightLeaf.Value...)...))
+
+	internalNodes := proof[2:]
+
+	for i := 0; i < len(internalNodes); i += 2 {
+		if internalNodes[i].Bitmask == leftChild {
+			// left sibling of the given node
+			hash = crypto.Hash(
+				append(append(internalNodes[i+1].Key, internalNodes[i+1].Value...),
+					append(internalNodes[i].Key, hash...)...),
+			)
+		} else {
+			// right sibling of the given node
+			hash = crypto.Hash(
+				append(append(internalNodes[i].Key, hash...),
+					append(internalNodes[i+1].Key, internalNodes[i+1].Value...)...),
+			)
+		}
+	}
+
+	// compare the calculated root hash against the provided root hash
+	if !bytes.Equal(hash, root) {
+		return false, nil
+	}
+
+	// Rebuild a similar merkle tree using the proof nodes, to be able to traverse it
+	// again and confirm whether the given key and value are part of the tree, to
+	// confirm the proof-of-non-membership as if the key is not part of the tree
+	memStore, err := NewStoreInMemory(lib.NewDefaultLogger())
+	if err != nil {
+		return false, err
+	}
+	smt := NewSMT(s.Root(), s.keyBitLength, memStore)
+
+	// add the nodes
+	for _, leaf := range proof {
+		// Keys are saved "as-is" to preserve the original values of the tree
+		// when the proof was obtained. Some intermediate node keys may have a length
+		// shorter than the MaxKeyBitLength. This pads such keys to ensure they are
```

Comment:
I'm doubtful that this is what we want to do here.

I think we want to write the keys 'as is' without padding

Reply:
Shouldn't we just use `setNode` for each?

```go
        // set the node in the database
	if err := s.setNode(s.target); err != nil {
		return err
	}
```

Reply:
I could try that, the original issue is that because some intermediate keys were made with a gcp lower than the keyBitLength, generating those keys would panic as the index created by keybitlength would be larger than the gcp length

Reply:
solved